[wasm-ep] Implement DiagnosticServer and startup sessions for WebAssembly

lambdageek opened the pull request:
Add a diagnostic server for WebAssembly.  Enable by building the runtime with `/p:WasmEnablePerfTracing=true` or `/p:WasmEnableThreads=true`.

To configure a project to start the diagnostic server, add this to the .csproj:
```xml
      <WasmExtraConfig Include="diagnostic_options" Value='
{
  "server": { "suspend": false, "connect_url": "ws://localhost:8088/diagnostics" }
}' />
```

The `connect_url` should be a WebSocket url serviced by `dotnet-dsrouter server-websocket` **from this branch** https://github.com/lambdageek/diagnostics/tree/wasm-server

Note that setting `"suspend": true` will hang the browser tab until a diagnostic tool such as `dotnet-trace collect` connects to the dsrouter.

---

Implement creating VFS file based sessions at runtime startup.  Add the following to a .csproj:

```xml
    <WasmExtraConfig Include="diagnostic_options" Value='
{
  "sessions": [ { "collectRundownEvents": "true", "providers": "WasmHello::5:EventCounterIntervalSec=1" } ]
}' />
```

That will create and start one or more EventPipe sessions that will store their results into the VFS.

The startup session can be retrieved via `MONO.diagnostics.getStartupSessions()`.  Each session `s` should be stopped via `s.stop()` and the data can then be extraced in a `Blob` using `s.getTraceBlob()`.

This is orthogonal to the diagnostic server support.  You don't need `dotnet-dsrouter` running on the host.  But you do need access to JavaScript on the main thread.

---

Notes/Issues:

* Tree shaking: I verified that if threads are not available, all the TypeScript diagnostics code is removed.
* Right now the server is not very robust to `dotnet-dsrouter` stopping, or starting after the runtime starts.  The ideal order is to start `dotnet-dsrouter` first, and then open the browser
* Unrelated housekeeping fixes:
   * Tell `wasm.proj` about all the subdirectories with .ts files - makes incremental builds notice changes in subdirectories.
   * Add a rollup `dependencies` property to quiet a warning about `node/buffer`
   * There's a mock implementation of a "websocket" that was used for protocol testing. I verified that tree-shaking removes this in thread-enabled Release builds.
   * Bump `PTHREAD_POOL_SIZE` to `4` and set `PTHREAD_POOL_SIZE_STRICT=2` (returns `EAGAIN` from `pthread_create` if the pool needs to grow).  The previous setting `PTHREAD_POOL_SIZE_STRING=1` (warn and try to grow the pool) seemed to lead to hangs.  Usually that means the main thread is creating a thread and immediately calling `pthread_join` without returning to JS. We should investigate separately.
   * The only implemented diagnostic server commands are `CollectTracing2`, `StopCollecting` and `ResumeRuntime`. None of the `Dump`, `Process` and `Profiler` commands are implemented and the server will crash if it receives them.  It should be relatively straightforward to return a "command unsupported" reply (which would allow clients to gracefully disconnect), but it's not done yet.
* In some error states the runtime kills the browser tab with the following in a terminal window (if Chrome is started from a terminal: `FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory`). This probably means we're hitting a loop somewhere that rapidly exhausts JIT memory, but it's difficult to investigate when the JS console dies, too (happens with chrome stable v103 and chrome beta v104).

Fixes https://github.com/dotnet/runtime/issues/69674, contributes to https://github.com/dotnet/runtime/issues/72481

lambdageek committed:
WIP: add a way to specify EP sessions in the MonoConfig

Currently not wired up to the runtime

lambdageek committed:
Add a mechanism to copy startup configs into the runtime and session
IDs out

lambdageek committed:
WIP: C side startup provider copying

lambdageek committed:
WIP checkpoint. Do more from JS

The issue is that once we're setting up streaming sessions, we will
need to send back a DS IPC reply with the session id before we start
streaming.  So it's better to just call back to JS when we start up
and setup all the EP sessions from JS so that when we return to C
everything is all ready.

lambdageek committed:
checkpoint: starting a session at startup works

lambdageek committed:
WIP: checkpoint add a controller and a webworker for DS

lambdageek committed:
WIP checkpoint EventPipeIPCSession class skeleton

lambdageek committed:
WIP checkpoint: runtime crashes; but WS from JS works

lambdageek committed:
XXX PrintfDebuggingHacks

lambdageek committed:
WIP: diagnostic server

lambdageek committed:
WIP some bits of the websocket worker

lambdageek committed:
WIP some notes on diagnostics and JS workers

lambdageek committed:
fix eslint

lambdageek committed:
debug printfs etc

more printfs

lambdageek committed:
[wasm] Enable the tracing component if threading is supported

lambdageek committed:
WIP: start moving the diagnostic server to a JS pthread

lambdageek committed:
WIP: move things around

lambdageek committed:
notes

lambdageek committed:
cleanup

lambdageek committed:
[diagnostic_server] wasm-specific fn_table

We won't be using the native C version

lambdageek committed:
[wasm-ep] disable DS connect ports in C, too

we will implement DS in JS

lambdageek committed:
asyncify finalize_startup; make 1 diagnostics init function

Combine mono_wasm_init_diagnostics and configure_diagnostics.
Now that finalize_startup is async, we can pause to wait for the
diagnostic server to startup at this point

lambdageek committed:
(not implemented) set browser-eventpipe sample to start a DS server

lambdageek committed:
ping in the DS server (not functional yet)

once we're able to start the DS server, make it log a ping to the console

lambdageek committed:
Start diagnostic server pthread

Clean up some of the old WIP code - we will probably not send
configuration strings from the diagnostic server back to the main thread.

lambdageek committed:
WIP try to start the server

It doesn't work right now because the MessagePort is not created until
the server thread attaches to Mono, which doesn't happen because it's
started before Mono.

Also it doesn't yet send a resume event, so the main thread just
blocks forever

lambdageek committed:
WIP diagnostic server server

lambdageek committed:
Add a mock WebSocket connection to simulate the remote end

Start the diagnostic server and have it perform the open/advertise
steps with the mock.

lambdageek committed:
cleanup diagnostics.ts

we're not going ot need a JS version of the DS EP streaming sessions.
And file-based auto-stop sessions are not going in

lambdageek committed:
wasm-mt: use a PThreadSelf struct instead of a raw MessagePort

lambdageek committed:
Move all the EP and diagnostic server modules to one directory

lambdageek committed:
Refactor; remove dead code; rationalize controller

the flow is now:

main -{creates pthread}->              server
  .                                      server creates event listener
  .  <-{sends diagnostic MessagePort}-   .
main creates event listener              .
  .  -{posts "start" message}->          .
  .                                      begins server loop

after the server loop is running, the main thread will get a
"resume_startup" message once the diagnostic server receives the right
command from the websocket.

next TODO: the runtime needs to send a "attach to runtime" message
which will signal the server that it can attach to the runtime (in
native) and start calling EP session creation functions.

lambdageek committed:
WIP more server pthread impl

lambdageek committed:
WIP: start adding queue from streaming thread to DS thread

We can't set up a shared MessagePort very easily (we need to bounce
through the main thread but it probably won't be able to process our
message until it's too late).

Also Atomics.waitAsync isn't available on many browsers (Chrome only).

So we use emscripten's dispatch mechanism to trigger an event in the
diagnostic thread to wake up and service the streaming thread's queue.
Right now the queue is dumb so we trigger on every write.  and also
the write is synchronous.

But it's simple to think about and it's implementable.

lambdageek committed:
[wasm] Incremental build and rollup warnings cleanups

- Add 'node/buffer' as an extrenal dependency. This doesn't do
anything except quiet a rollup warning about the import.
- Add all the .ts files, and the tsconfig files (except node_modules)
to the rollup inputs, to make sure we re-run rollup when anything changes.

lambdageek committed:
WIP: work on wiring up DS protocol commands (mock); resume hack

- start adding commands so that we can strt some sessions from DS
- we can't avoid a busy loop in ds_server_wasm_pause_for_diagnostics_monitor.
  we can't make the main thread pause until we get a resume command
  until after we're able to start an EP session (DS client won't send
  a resume command until we send an EP session ID back).  If the DS
  pauses until it can attach to the runtime, and the runtime pauses
  until DS tells it to resume, the main thread pause has to be after
  we get EP and DS initialized.  But that means it can't be async.  So
  we'll just have to busy wait on a condition variable in native.

lambdageek committed:
WIP: set up a WasmIpcStream, create EP sessions from DS

Seems to create the sesssion, but not seeing write events
yet. possibly due to not flushing?

lambdageek committed:
WIP: starting to stream works; needs PTHREAD_POOL_SIZE bump

Looks like we can send the initial nettrace header some events.

We're starting more threads, so we need a bigger thread pool.

Also PTHREAD_POOL_SIZE_STRICT=1 (the default - warn if worker pool needs to grow,
but still try to grow it) seems to deadlock the browser-eventpipe
sample.

Set PTHREAD_POOL_SIZE_STRICT=2 (don't try to allocate a worker, make
pthread_create fail with EAGAIN) instead so we get some kind of
exception instead in other circumstances.

Set the pool size to 4.

lambdageek committed:
cleanup browser-eventpipe sample

lambdageek committed:
refactor to simplify and cleanup; rm duplicate code

lambdageek committed:
call mono_wasm_event_pipe_early_startup_callback from event_pipe init

instead of from the rundown_execution_checkpoint_2 function

lambdageek committed:
if diagnostics server isn't enabled, don't try to initialize it

lambdageek committed:
WIP: start parsing binary commands

lambdageek committed:
WIP: start wiring up binary protocol parsing to the websocket

lambdageek committed:
WIP: Can parse a CollectTracing2 command and attempt to create a
session!

lambdageek committed:
[wasm-ep] use the new PromiseController<T>

lambdageek committed:
get back to the server loop quicker by queueing the parsing in the microtask

lambdageek committed:
update mock for binary ADVR_V1 message

lambdageek committed:
sample: don't suspend, and use a mock url

lambdageek committed:
use better values for parse results

lambdageek committed:
parse a few more binary protocol commands

lambdageek committed:
wasm_ipc_stream: wire up close command

Use a sentinal "buf" value (-1) to signal that the writer closed the stream

lambdageek committed:
Send proper OK messages in replies to binary protocol commands

lambdageek committed:
(testing) turn off the file session for now

lambdageek committed:
TODO: handle WS connection failures

lambdageek committed:
remove em_asm(console.log); simplify wasm EP init

Just call the EP JS callback directly from native

lambdageek committed:
remove debug output

lambdageek committed:
remove debug output in startup

lambdageek committed:
cleanup wasm ipc stream impl

lambdageek committed:
put diagnostics mocks behind a const flag

lambdageek committed:
don't build wasm-specific DS if threads are disabled

lambdageek committed:
refactor and cleanup

- Move the IPC parsing and serialization into separate files
- Try to have one responsibility per class
- update comments and docs

lambdageek committed:
help treeshaking

verified that all the DS and EP JS code is dropped if monoWasmThreads
is falsed.

lambdageek committed:
update DS design notes

lambdageek committed:
remove more printfs

lambdageek committed:
use PromiseController in more places

lambdageek committed:
remove more console.debug in startup

lambdageek committed:
Merge remote-tracking branch 'origin/main' into wasm-ep-on-startup

lambdageek commented:
@pavelsavara this is going to have a merge conflict with #72275 in `startup.ts` (and probably other minor ones).
The main thing that I changed was to make `_apply_configuration_from_args` into an async function (so that I can call the async `mono_wasm_diagnostics_init` which needs to make the main thread wait until the diagnostic server thread starts.
There is probably some way to do the same thing in the new setup?  Is it possible to make `mono_wasm_before_user_runtime_initialized` async?

lambdageek committed:
add MONO_WASM prefix to console logging outputs

lambdageek committed:
fix Windows build

pavelsavara commented on src/mono/wasm/runtime/workers/README.md:
```diff
@@ -11,6 +11,17 @@ file (or `.js`) and a `xyz/` subdirectory with additional sources.
+
+- create `xyz/shared/`, `xyz/browser/` and `xyz/worker/` directories that have `index.ts` and `tsconfig.json` files that are set up for common ES, ES with DOM APIs and ES with WebWorker APIs, respectively
+- call the apropriate functions (browser or worker) from the C code or from JS.
+
```
Could you please describe how do you enable the mocks and how to use them for testing ? Is there unit test which uses it ?

lambdageek replied:
Just to update the status: I'm working on a couple of changes to the mock stuff:
1. Move the actual mock protocol script into the sample project.  The plan is is that the .csproj `diagnostic_options.server_url` json can be `mock:relative_url/mock.js` and then we will `import(mock_url).then(module => mock(module.default).run()` - ie: the mock script is per-project and it's just an es6 module with a default export in a format that the mock engine understands.
2. I'm adding some utilities to the mock engine so it can say things like `connection.expectOkReply ((payload) => /*do a check of the payload */)` rather than leaving the mock script to parse the bytes on its own.
3. document how to use it.

Should be done by EOD today.

lambdageek replied:
ok, mocking changes are up

lambdageek replied:
I also generate `diagnostics-mock.d.ts` into the `wasm/runtime` directory in debug builds, same as `dotnet.d.ts`. I found it helpful to have TypeScript check my JS mocking script.

pavelsavara commented:
> I tested merging the branches. Looks like the only thing I had to do was to make `mono_wasm_before_user_runtime_initialized` async.

Good, I don't know if I would merge my PR in Net7, probably just smaller portion of it. So feel free to merge as is.
Looked at the typescript and LGTM.
It would be good to have JavaScript-only unit tests for complex components at some point. https://jestjs.io/ is good choice I think.

lateralusX commented on src/mono/mono/component/event_pipe.c:
```diff
@@ -354,6 +363,7 @@ wasm_to_ep_session_id (MonoWasmEventPipeSessionID session_id)
 
 EMSCRIPTEN_KEEPALIVE gboolean
 mono_wasm_event_pipe_enable (const ep_char8_t *output_path,
+			     IpcStream *ipc_stream,
```
Remove IpcStream comment below, since it has been moved.

lateralusX reviewed (commented):
Native EventPipe bits LGTM!

lambdageek commented on src/mono/mono/component/diagnostics_server.c:
```diff
@@ -19,6 +28,291 @@ static MonoComponentDiagnosticsServer fn_table = {
+				break;
+			else {
+				/* timed out */
+				cumulative_timeout += timeout;
```
I used to print a warning to the JS console here, but it's not actually that useful yet because it's difficult to access the JS console while the browser tab is busy.

I'll simplify this for now.

lambdageek commented on src/mono/mono.proj:
```diff
@@ -483,6 +483,7 @@
     <ItemGroup Condition="'$(TargetsBrowser)' == 'true'">
       <_MonoCMakeArgs Include="-DFEATURE_PERFTRACING_DISABLE_PERFTRACING_LISTEN_PORTS=1"/>
       <_MonoCMakeArgs Include="-DFEATURE_PERFTRACING_DISABLE_DEFAULT_LISTEN_PORT=1"/>
+      <_MonoCMakeArgs Include="-DFEATURE_PERFTRACING_DISABLE_CONNECT_PORTS=1" />
```
yea the idea was to disable most of the C diagnostic server functionality as it is not needed.

lambdageek committed:
fix sample logic when startup session is disabled

lambdageek committed:
improve debug output for DS server

keep track of open/advertise counts and print them when receiving replies

lambdageek committed:
bugfix: don't confuse buf_addr for the value stored in it

the buf_addr is always the same for a given queue. the value in it is
what we need to check to see if it's the sentinel value

lambdageek committed:
slight refactor of EventPipeSocketConnection and more logging

lambdageek committed:
review feedback

lambdageek committed:
Merge remote-tracking branch 'origin/main' into wasm-ep-on-startup

lambdageek committed:
merge fixup

lambdageek committed:
fix bug in queue_push_sync main thread detection

lambdageek committed:
fix typo

lambdageek committed:
make diagnostic server mocking friendlier

Allow each test project to specify its own mock script.

Also provide TypeScript declarations for the mocking interfaces

Also always use binary protocol commands - don't send json for mocks.

lambdageek committed:
Merge remote-tracking branch 'origin/main' into wasm-ep-on-startup

lambdageek committed:
merge fixup

lambdageek committed:
fix rollup warning when making the crypto worker

pavelsavara commented on src/mono/wasm/runtime/startup.ts:
```diff
@@ -788,8 +793,7 @@ export type DownloadAssetsContext = {
-    console.debug("mono_wasm_pthread_worker_init");
     pthreads_worker.currentWorkerThreadEvents.addEventListener(pthreads_worker.dotnetPthreadCreated, (ev) => {
-        console.debug("thread created", ev.pthread_ptr);
+        console.debug("pthread created", ev.pthread_self.pthread_id);
```
`MONO_WASM: ` prefix

lambdageek commented:
> It would be good to have JavaScript-only unit tests for complex components at some point. https://jestjs.io/ is good choice I think.

I'll play around with that this afternoon.  Probably won't be part of this PR unless it turns out to be completely trivial to add the unit tests to our setup.

lambdageek committed:
add MONO_WASM: prefix to logging

lambdageek committed:
disable mocking in the sample project by default

lambdageek commented:
verified one more time that with threading or perftracing disabled, all of the related JS is removed by tree-shaking

lambdageek committed:
Merge remote-tracking branch 'origin/main' into wasm-ep-on-startup

lambdageek committed:
fixup after merge

pavelsavara commented on src/mono/wasm/runtime/diagnostics/README.md:
```diff
@@ -0,0 +1,92 @@
+- `shared/` type definitions to be shared between the worker and browser main thread
+- `mock/` a utility to fake WebSocket connectings by playing back a script.  Used for prototyping the diagnostic server without hooking up to a real WebSocket.
+
+## Mocking
```
What is it we are mocking and why ? (high level intro)

pavelsavara replied:
Do I understand that the diagnosed process is a "server" and the tooling outside of the browser is a "client" ? And so, we are mocking the client here and it's requests ?

pavelsavara commented on src/mono/wasm/runtime/diagnostics/shared/create-session.ts:
```diff
@@ -0,0 +1,52 @@
+    if (!cwraps.mono_wasm_event_pipe_enable(tracePath, ipcStreamAddr, options.bufferSizeInMB, options.providers, options.rundownRequested, sessionIdOutPtr)) {
+        return false;
+    } else {
+        return memory.getI32(sessionIdOutPtr);
```
getU32 for pointers

pavelsavara commented on src/mono/wasm/runtime/polyfills.ts:
```diff
@@ -41,47 +41,66 @@ export async function init_polyfills(): Promise<void> {
+                for (const sub of subscribers) {
+                    const listener = sub.listener;
+                    if (sub.oneShot) {
+                        this.removeEventListener(event.type, listener);
```
do we need copy of `subscribers` because of the change in the loop ?

pavelsavara commented on src/mono/wasm/runtime/pthreads/worker/index.ts:
```diff
@@ -23,40 +25,57 @@ export {
 
 function monoDedicatedChannelMessageFromMainToWorker(event: MessageEvent<string>): void {
-    console.debug("got message from main on the dedicated channel", event.data);
+    console.debug("MONO_WASM: got message from main on the dedicated channel", event.data);
```
After we merge both this and my PR, we could prefix this with `if (runtimeHelpers.diagnostic_tracing)` so that by default the runtime is not noisy. It matters in console like nodeJS. @maraf suggested it's about a time when we introduce logging component to centralize all this logic to one place.

lambdageek replied:
agreed

pavelsavara reviewed (approved):
LGTM, none of my other comments are blocking merge.

lambdageek committed:
review feedback

- improve diagnostics mock README
- note that mocking just uses ES6 modules, testing with CJS is not
   supported right now.
- fix iteration over listeners when dispatching a one-shot event in
  the EventTargt polyfill
- use U32 getter in EP session creation

lambdageek commented on src/mono/wasm/runtime/diagnostics/server_pthread/mock-remote.ts:
```diff
@@ -0,0 +1,20 @@
+    if (monoDiagnosticsMock) {
+        const mockPrefix = "mock:";
+        const scriptURL = mockURL.substring(mockPrefix.length);
+        return import(scriptURL).then((mockModule) => {
```
Yea, if there's a need for writing the mocking script using CJS, we can add it here.  But since the goal here is just to have a tool to improve the server's robustness (for example: make scripts that send garbage and verify that the DS can recover), I don't think there's a need for CJS.  This isn't something we're ever going to ship. It's just for better testing.

lambdageek commented:
https://helix.dot.net/api/2019-06-17/jobs/d29bfb53-4790-47f4-8a1e-aa868ec7afd5/workitems/System.Net.Mail.Functional.Tests/console failure is unrelated

```
/__w/1/s/src/native/libs/Common/pal_utilities.h:86: int ToFileDescriptor(intptr_t): Assertion `0 <= fd && fd < sysconf(_SC_OPEN_MAX)' failed.
```

pavelsavara commented:
/azp run runtime-wasm
